### PR TITLE
openapi-aggregation: speed up merging from 1 sec to 50-100 ms

### PR DESCRIPTION
**Pull request by sttts**
This PR avoids swagger spec cloning by non-source-mutating merging. It uses a mutating walker that locally clones minimal substructures.

Speedup: >>10

```release-note
Considerably reduced the CPU load in kube-apiserver while aggregating OpenAPI specifications from aggregated API servers.
```

### CODE REVIEW

**Comment by roycaihw** [comment]
Great! The idea looks good to me. Spec cloning makes up ~98% of the time when we do spec merging. 

We need to unit test the cloning. I see kube-openapi has a decent amount of [unit test](https://github.com/kubernetes/kube-openapi/blob/master/pkg/aggregator/aggregator_test.go) covering spec merging. Maybe that's enough

File: vendor/k8s.io/kube-openapi/pkg/aggregator/mutating_walker.go [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+	cloned := false
+	clone := func() {
+		if !cloned {
+			cloned = true
+			swagger := &spec.Swagger{}
+			*swagger = *orig
+		}
+	}
+
+	ParametersCloned := false
```
Comment (juanvallejo) [inline vendor/k8s.io/kube-openapi/pkg/aggregator/mutating_walker.go]: does this var need to be exported?

File: vendor/k8s.io/kube-openapi/pkg/aggregator/mutating_walker.go [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+}
+
+func (w *mutatingReferenceWalker) walkSchema(schema *spec.Schema) *spec.Schema {
+	if schema == nil {
+		return nil
+	}
+
+	orig := schema
+	clone := func() {
+		if orig == schema {
```
[the remark this answers is not in the document]
Comment (sttts) [reply vendor/k8s.io/kube-openapi/pkg/aggregator/mutating_walker.go]: 40x speedup, 27x less allocs.

File: vendor/k8s.io/kube-openapi/pkg/aggregator/mutating_walker.go [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+	for k, v := range schema.Definitions {
+		s := w.walkSchema(&v)
+		if s != &v {
+			if !definitionsCloned {
+				definitionsCloned = true
+				clone()
+				schema.Definitions = make(spec.Definitions, len(orig.Definitions))
+				for k2, v2 := range orig.Definitions {
+					if k2 == k {
+						break
```
[the remark this answers is not in the document]
Comment (sttts) [reply vendor/k8s.io/kube-openapi/pkg/aggregator/mutating_walker.go]: this was fixed

**Comment by mbohlool** [comment]
This is a change in vendor folder, please close this in favor of the other PR for less confusion. I was looking for my comments on the other PR here :)

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go [hunk 4]
```diff
@@ -83,6 +84,16 @@ func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.
 		return nil, err
 	}
 
+	// roundtrip aggregator spec because it is the only one which does not go through the delegate HTTP handler.
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go]: this is hard to follow... why is the spec returned from BuildOpenAPISpec unsuitable? can we fix the output of that function?

File: hack/update-openapi-spec.sh [hunk 5]
```diff
@@ -83,7 +84,7 @@ fi
 
 kube::log::status "Updating " ${OPENAPI_ROOT_DIR}
 
-curl -w "\n" -fs "${API_HOST}:${API_PORT}/openapi/v2" > "${OPENAPI_ROOT_DIR}/swagger.json"
+curl -w "\n" -fs "${API_HOST}:${API_PORT}/openapi/v2" | jq . > "${OPENAPI_ROOT_DIR}/swagger.json"
```
Comment (liggitt) [inline hack/update-openapi-spec.sh]: this unescapes `<` and `>` chars in output... not necessarily a problem, just noting it

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go]: The aggregator spec had non-nil, empty `Scheme.Properties`. After roundtrip these turned into `nil`s for all other delegated apiserver specs (those we download via HTTP handlers, yes crazy, including JSON unmarshaling, at least with etags (I hope); different topic). The consequence is that the merge algorithm thinks the definitions are different, which leads to renames and duplication of all meta/v1 structs.

File: hack/update-openapi-spec.sh (hunk 5 above)
Comment (sttts) [reply hack/update-openapi-spec.sh]: I this file more than documentation? Where is it actually used?

Comment (liggitt) [reply hack/update-openapi-spec.sh]: it gets loaded as a fixture by some kubectl tests, otherwise mostly as documentation

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go]: I added a [quick unit test](https://gist.github.com/liggitt/37c3d3b798bf0b392f389dd74458890b#file-openapi_test-go) to see what the diff between in-memory and round-tripped schemas were, and the [results](https://gist.github.com/liggitt/37c3d3b798bf0b392f389dd74458890b#file-openapi_test_results-txt) were scary. In addition to the nil vs `[]`/`{}` example, it looks like additionalProperties is opposite in memory and once round tripped.

Rather than ad-hoc fixup here, can we adjust the generator to produce things that roundtrip without changes?

File: hack/update-openapi-spec.sh (hunk 5 above)
Comment (roycaihw) [reply hack/update-openapi-spec.sh]: the checked-in spec is also used for generating [clients](https://github.com/kubernetes-client). the generator loads the spec from release branch

Comment (liggitt) [reply hack/update-openapi-spec.sh]: To be clear, it's still valid JSON and a valid spec, just doesn't have HTML-escaped angle brackets

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go]: what does "opposite" mean here?

> can we adjust the generator to produce things that roundtrip without changes?

I don't think is possible. Vendor extensions are just `interface{}`s. There is no scheme like in kube which could translate them back to native structs, you will just get some JSON blob instead.

Comment (liggitt) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go]: > what does "opposite" mean here?

in-memory, additionalProperties is `false`, which gets omitted when serialized, then defaults to `true` when deserialized

**Comment by liggitt** [comment]
something is up with the godeps changes:
```
Your Godeps.json is different:
--- Godeps/Godeps.json	2019-02-08 20:33:07.190360977 +0000
+++ /tmp/gopath.tGCkZA/src/k8s.io/kubernetes/Godeps/Godeps.json	2019-02-08 20:47:27.121379369 +0000
@@ -1679,7 +1679,7 @@
 		},
 		{
 			"ImportPath": "github.com/go-sql-driver/mysql",
-			"Comment": "v1.3",
+			"Comment": "v1.3.0",
 			"Rev": "a0583e0143b1624142adab07e0e97fe106d99561"
 		},
 		{
@@ -2348,6 +2348,7 @@
 		},
 		{
 			"ImportPath": "github.com/inconshreveable/mousetrap",
+			"Comment": "v1.0",
 			"Rev": "76626ae9c91c4f2a10f34cad8ce83ea42c93bb75"
 		},
 		{
@@ -2429,12 +2430,10 @@
 		},
 		{
 			"ImportPath": "github.com/lib/pq",
-			"Comment": "go1.0-cutoff-203-g88edab0803230a",
 			"Rev": "88edab0803230a3898347e77b474f8c1820a1f20"
 		},
 		{
 			"ImportPath": "github.com/lib/pq/oid",
-			"Comment": "go1.0-cutoff-203-g88edab0803230a",
 			"Rev": "88edab0803230a3898347e77b474f8c1820a1f20"
 		},
 		{
@@ -2508,7 +2507,7 @@
 		},
 		{
 			"ImportPath": "github.com/matttproud/golang_protobuf_extensions/pbutil",
-			"Comment": "v1.0.0-2-gc12348ce28de40",
+			"Comment": "v1.0.1",
 			"Rev": "c12348ce28de40eed0136aa2b644d0ee0650e56c"
 		},
```

File: Godeps/Godeps.json [hunk 6]
```diff
@@ -1679,7 +1679,7 @@
 		},
 		{
 			"ImportPath": "github.com/go-sql-driver/mysql",
-			"Comment": "v1.3.0",
+			"Comment": "v1.3",
```
Comment (sttts) [inline Godeps/Godeps.json]: @cblecker these comment changes are ok? Or anything broken?

Comment (sttts) [reply Godeps/Godeps.json]: for some reason godep-restore does not download all tags here. If I run `git fetch` in each repo, godep-save is happy.

**Comment by liggitt** [comment]
/retest

verified the only differences in the swagger.json file are object key order and json string escaping (no structural or value changes that would affect anything consuming the published schemas)

```sh
diff \
  <(curl -s https://raw.githubusercontent.com/sttts/kubernetes/sttts-openapi-aggreation-without-clone/api/openapi-spec/swagger.json | jq -S .) \
  <(curl -s https://raw.githubusercontent.com/kubernetes/kubernetes/master/api/openapi-spec/swagger.json | jq -S .)

<no diff>
```

/lgtm
/approve
/retest
